pkg/odo/genericclioptions/context.go

```diff
@@ -31,15 +37,99 @@ type Context struct {
 // internalCxt holds the actual context values and is not exported so that it cannot be instantiated outside of this package.
 // This ensures that Context objects are always created properly via NewContext factory functions.
 type internalCxt struct {
-	Client          *occlient.Client
-	command         *cobra.Command
-	Project         string
-	Application     string
-	cmp             string
-	OutputFlag      string
-	LocalConfigInfo *config.LocalConfigInfo
-	KClient         *kclient.Client
-	EnvSpecificInfo *envinfo.EnvSpecificInfo
+	ComponentContext    string
+	Client              *occlient.Client
+	command             *cobra.Command
+	Project             string
+	Application         string
+	cmp                 string
+	OutputFlag          string
+	LocalConfigInfo     *config.LocalConfigInfo
+	KClient             *kclient.Client
+	EnvSpecificInfo     *envinfo.EnvSpecificInfo
+	LocalConfigProvider localConfigProvider.LocalConfigProvider
+}
+
+type ContextOptions struct {
+	Cmd              *cobra.Command
+	DevfilePath      string
+	ComponentContext string
+	IsNow            bool
+}
+
+func New(options ContextOptions, toggles ...bool) (*Context, error) {
+	options.DevfilePath = completeDevfilePath(options.ComponentContext, options.DevfilePath)
+
+	var context *Context
+	isDevfile := odoutil.CheckPathExists(options.DevfilePath)
+	if isDevfile {
+		context = NewDevfileContext(options.Cmd)
+		context.ComponentContext = options.ComponentContext
+
+		err := context.InitEnvInfoFromContext()
+		if err != nil {
+			return nil, err
+		}
+
+		// Parse devfile and validate
```

Comment:
We are parsing and validating here because we need to use it in the envinfo and we are also setting the value of the localConfigProvider later in the function.

Reply:
OK. But what part of parsing and validation is used in setting `EnvInfo` as the config provider? As far as I can see, it's only this call:
```go
context.EnvSpecificInfo.SetDevfileObj(devObj)
```
that's using the output of parsing & validation. But the question is that is this really the right place to set DevfileObj or should this be done elsewhere? 

This is a genuine question. I'm just trying to make sure we're not doing something that can safely (or, actually) be done later.

Reply:
> But the question is that is this really the right place to set DevfileObj or should this be done elsewhere?

Yes because most of the other packages shouldn't know about the existence of an envInfo and localConfig

Reply:
Makes sense. Thanks. :+1: